src/NHibernate/Engine/Loading/CollectionLoadContext.cs

```diff
@@ -145,7 +145,20 @@ public IPersistentCollection GetLoadingCollection(ICollectionPersister persister
 		/// complete. 
 		/// </summary>
 		/// <param name="persister">The persister for which to complete loading. </param>
+		[Obsolete("Please use EndLoadingCollections(ICollectionPersister, HashSet<string>) instead.")]
 		public void EndLoadingCollections(ICollectionPersister persister)
+		{
+			EndLoadingCollections(persister, null);
+		}
+
+		/// <summary> 
+		/// Finish the process of collection-loading for this bound result set.  Mainly this
+		/// involves cleaning up resources and notifying the collections that loading is
+		/// complete. 
+		/// </summary>
+		/// <param name="persister">The persister for which to complete loading. </param>
+		/// <param name="uncacheableCollections">Indicates if collections can be put in cache</param>
+		public void EndLoadingCollections(ICollectionPersister persister, HashSet<string> uncacheableCollections)
```

Comment:
I think here we can simplify signature to provide bool flag instead of HashSet:
`public void EndLoadingCollections(ICollectionPersister persister, bool skipCache)`

Reply:
It seems initially it was only a `bool` all the way long, but on the JIRA:

>**Alexander Zaytsev** added a comment — <time datetime="2017-04-07T01:37:44.330Z" >title="2017-04-07T01:37:44.330Z">7th April 2017, 1:37:44</time>:
>> <~Lukasz> your pull request(s) do not consider 2 separate fetches. When there is a collection *with* restrictions and another one *without* then I expect that the collection *without* restrictions would be cached.
>> 
>> So, the property "CanAddCollectionsToCache" shall probably be a dictionary.
>> 

But well, I am not valuing much handling this case, because two collection fetches causes a Cartesian product, so in most cases it should be avoided.

Reply:
@fredericDelaporte I'm not sure it's related to this particular place. From what I see here we provide exact collection `persister` to fetch.  So for multiple fetched collections it's multiple calls and it's up to caller to calculate `skipCache` flag for this particular `persister`. So with my suggestion it's still possible to mix cached and uncached collections.

To be more precise I think we can have in `Loader` optional array 
`protected virtual bool[] UncachedCollectionPersisters`

And check it something like:
`EndCollectionLoad(resultSetId, session, collectionPersisters[i], skipCache: UncachedCollectionPersisters?[i] == true);`